### PR TITLE
Fix bug in MergeJoin with low cardinality in using

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Fix possible crash in `partial_merge` and `auto` join algorithms in with `USING` on `LowCardinality` column.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 56714928f6b9546759f63610e4204addbc69f4dc with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53125/56714928f6b9546759f63610e4204addbc69f4dc/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Interpreters/MergeJoin.cpp [hunk 1]
```diff
@@ -532,16 +532,31 @@ MergeJoin::MergeJoin(std::shared_ptr<TableJoin> table_join_, const Block & right
...
+        if (table_join->hasUsing() && !required_right_keys.contains(right_key))
+        {
+            /*
+             * If the column from the 'USING' clause has low cardinality only on the right-hand side,
+             * we would not need to keep its low cardinality.
+             * Example:
+             *   SELECT a FROM (SELECT 1 :: LowCardinality(UInt32) AS a) t1
+             *   JOIN (SELECT 1 :: UInt32 AS a) t2 USING (a);
+             * In that case result `a` will be supertype of `t1.a` and `t2.a`, just `UInt32`
+             * If column in the left hand side is also lowcardinality, we will restore propery in joinBlock.
```
Comment (vdimir) [inline src/Interpreters/MergeJoin.cpp:547]:
```suggestion
             * If column in the left hand side is also lowcardinality, we will restore property in joinBlock.
```

File: src/Interpreters/MergeJoin.cpp [hunk 2]
```diff
@@ -532,16 +532,31 @@ MergeJoin::MergeJoin(std::shared_ptr<TableJoin> table_join_, const Block & right
     addConditionJoinColumn(right_sample_block, JoinTableSide::Right);
     JoinCommon::splitAdditionalColumns(key_names_right, right_sample_block, right_table_keys, right_columns_to_add);
 
+    const NameSet required_right_keys = table_join->requiredRightKeys();
     for (const auto & right_key : key_names_right)
     {
+        if (table_join->hasUsing() && !required_right_keys.contains(right_key))
```
Comment (novikd) [inline src/Interpreters/MergeJoin.cpp:538]: It's quite unclear why it checks for `LowCardinality`

Comment (vdimir) [reply src/Interpreters/MergeJoin.cpp:538]: We convert every lowcardinality key column to full inside MergeJoin. But in order to have the same types in output we restore lowcardinality back. For that we save such key names in `lowcard_right_keys`. The issue is that to case described below in a comment we need to have non-lowcardinality type in output, but return lowcard.

**Review by novikd — APPROVED** [review]

**Pull request closed without merging**